lib/multigrid.cpp

```diff
@@ -1429,4 +1387,53 @@ namespace quda
     popLevel(param.level);
   }
 
+  void MG::generateEigenVectors()
+  {
+    pushLevel(param.level);
+
+    // Extract eigensolver params from the size of the null space.
+    int nConv = param.B.size();
+    param.mg_global.eig_param[param.level]->nConv = nConv;
+    int nEv = param.B.size() + 10;
+    param.mg_global.eig_param[param.level]->nEv = nEv;
+    // int nKr = nEv + nEv / 2;
+    int nKr = 2 * nEv;
+    param.mg_global.eig_param[param.level]->nKr = nKr;
+
+    // Dummy array to keep the eigensolver happy.
+    std::vector<Complex> evals(nEv, 0.0);
+
+    std::vector<ColorSpinorField *> B_evecs;
+    ColorSpinorParam csParam(*param.B[0]);
+    csParam.gammaBasis = QUDA_UKQCD_GAMMA_BASIS;
+    csParam.create = QUDA_ZERO_FIELD_CREATE;
+    // This is the vector precision used by matResidual
+    csParam.setPrecision(param.mg_global.invert_param->cuda_prec_sloppy, QUDA_INVALID_PRECISION, true);
+
+    for (int i = 0; i < nKr; i++) B_evecs.push_back(ColorSpinorField::Create(csParam));
+
+    // before entering the eigen solver, lets free the B vectors to save some memory
+    ColorSpinorParam bParam(*param.B[0]);
+    for (int i = 0; i < (int)param.B.size(); i++) delete param.B[i];
+
+    DiracMdagM *mdagm = new DiracMdagM(*diracResidual);
```

Comment:
Do we want to expose the option of using `DiracMMdag` instead? (In practice I don't think it makes sense to use left singular vectors, so this isn't a hard request.)

Reply:
If it doesn't make sense, then I won't expose the option. We use Lanczos here not to extract any particular information about the spectrum of a particular operator (as we would in the stand alone solver) but to simply deflate a CG solve.

Reply:
It does make sense to expose the option.  When doing CGNE, you solve using MMdag not MMdag.

Reply:
Noting this is fixed now